keps/sig-instrumentation/20190425-metrics-watch-api.md

```diff
@@ -0,0 +1,191 @@
+authors:
+  - "@x13n"
+owning-sig: sig-instrumentation
+participating-sigs:
+  - sig-autoscaling
+  - (?)
+reviewers:
+  - TBD
+approvers:
+  - TBD
+editor: TBD
+creation-date: 2019-04-25
+last-updated: 2019-04-29
+status: provisional
+see-also: []
+replaces: []
+superseded-by: []
+---
+
+# Metrics API watch support
+
+## Table of Contents
+
+ * [Metrics API watch support](#metrics-api-watch-support)
+    * [Table of Contents](#table-of-contents)
+    * [Summary](#summary)
+    * [Motivation](#motivation)
+       * [Goals](#goals)
+       * [Non-Goals](#non-goals)
+    * [Proposal](#proposal)
+       * [User Stories](#user-stories)
+          * [HPA](#hpa)
+          * [Custom metrics provider](#custom-metrics-provider)
+       * [Implementation Details/Notes/Constraints [optional]](#implementation-detailsnotesconstraints-optional)
+       * [Risks and Mitigations](#risks-and-mitigations)
+    * [Design Details](#design-details)
+       * [Test Plan](#test-plan)
+       * [Graduation Criteria](#graduation-criteria)
+       * [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+       * [Version Skew Strategy](#version-skew-strategy)
+    * [Implementation History](#implementation-history)
+    * [Drawbacks [optional]](#drawbacks-optional)
+ * [Related resources](#related-resources)
+
+## Summary
+
+Provide watch capability to all resource metrics APIs: `metrics.k8s.io`,
+`custom.metrics.k8s.io` and `external.metrics.k8s.io`, [similarly to regular
+Kubernetes APIs](https://kubernetes.io/docs/reference/using-api/api-concepts/#efficient-detection-of-changes).
+
+## Motivation
+
+There are three APIs for serving metrics. All of them support reading the
+metrics in a request-response manner, forcing the clients (e.g. HPA) interested
+in up-to-date values to poll in a loop. This introduces additional latency:
+between the time when new metric values are known to the process serving the
+API and the time when a client interested in reading them actually fetches the
+data.
+
+### Goals
+
+- Allow resource metrics clients to subscribe to stream metric changes.
+
+### Non-Goals
+
+- Graduate the APIs to GA. This needs to be done eventually, but is out of scope
+  for this work.
+
+## Proposal
+
+This proposal is essentially about implementing [Efficient detection of
+changes](https://kubernetes.io/docs/reference/using-api/api-concepts/#efficient-detection-of-changes)
+for metrics. GET requests will accept additional `watch` parameter, which would
+cause the API to start streaming results. Since old metric values are never
+modified, the only supported update type will be `ADDED`, when a new data point
+appears. Metrics don't contain any resourceVersion associated with them, so it
+won't be possible to retrieve old values by passing `resourceVersion` parameter.
+Instead, this parameter will be ignored and all recent data points will be
+returned instead. This means metrics APIs will never return `410 Gone` error
```

Comment:
Wait - i think that either the description here is unclear (at least to me) or I see a significant issues with what was described here:

1. Maybe I'm missing something, but what do you mean by "old metric values are never modified"?
While conceptually this is true, the API objects are around Pod, Node, etc.
So actually, the object (unless I'm signiifcantly missing something) actually changes.
I actually believe that sending ADDED event on every event is wrong - it's simply not what it should be (and it would also break all the libraries build, like reflector/informer framework).
2. Conceptually, what you really implement here is the existing "resourceVersion=0" semantic of watch:
https://github.com/kubernetes/kubernetes/issues/13969
If that's really the case, we should simply make it explicit, and support only watch with RV=0 

The high level comment is that we shouldn't reuse the existing options/fields and give them different semantics than in all other places - it would be extremely misleading for users.

@kubernetes/sig-api-machinery-feature-requests @kubernetes/sig-api-machinery-api-reviews @jpbetz 
I think this is something that requires much closer collaboration wich apimachinery team.

Reply:
We only went ahead with merging this, as it’s provisional, not accepted, this is a work in progress state, but generally watch is desirable. I agree this needs very very close collaboration with apimachinery and autoscaling, I’ve mentioned this to the authors repeatedly :)

Reply:
@liggitt, @deads2k, or myself should probably make a pass over this.

Reply:
The fact that they don't use RV doesn't change anything - if you don't support it, you can validate the request and just support RV=0 (or sth like that).
But if the metrics for a given Pod or Node change, what you get is the same object (it has the same name, namespace, etc., right?), not a new object (so you can't send ADD, without sending DELETE for the previous one first).

2. I don't know - I would like to see it removed, but I'm afraid we won't have manpower to do it anytime soon.